proposals/0048-container-api.md

```diff
@@ -0,0 +1,196 @@
+- Start Date: 2024-04-26
+- Reference Issues: https://github.com/withastro/roadmap/issues/533
+- Implementation PR: 
+
+# Summary
+
+Astro components are tightly coupled to `Astro` (the metaframework). This proposal introduces a possible server-side API for rendering `.astro` files in isolation, outside the full `astro` build pipeline.
+
+# Example
+
+```js
+import { AstroContainer } from 'astro/container';
+import Component from './components/Component.astro';
+
+const container = await AstroContainer.create();
+console.log(await container.renderToString(Component, { props, slots }))
+```
+
+The container can be optionally constructed with some settings that are typically reserved for Astro configuration.
+```js
+const container = await AstroContainer.create({
+    mode: "production",
+    streaming: true,
+    astroConfig: {
+        site: "https://example.com",
+        trailingSlash: false
+    }
+})
+```
+
+The second argument to `renderToString` optionally provides render-specific data that may be exposed through the Astro global, like `props`, `slots`, `request`, and/or `params`.
+```js
+await astro.renderToString(Component, { props, slots, request, params })
+```
+
+# Background & Motivation
+
+Some of our own proposals, as well as many third-party tool integrations, are blocked until we expose a proper server-side rendering API for `.astro` components. Other frameworks have tools like `preact-render-to-string` or `react-dom/server` that make rendering components on the server straight-forward.
+
+We've avoided doing this because... it's very hard! Part of the appeal of `.astro` components is that they have immediate access to a number of powerful APIs, context about your site, and a rich set of framework renderers. In the full `astro` build process, we are able to wire up all this context invisibly. Wiring up the context manually as a third-party is prohibitively complex.
+
+
+# Goals
+
+- Provide a **low-level** API for rendering `.astro` components in isolation
+- Expose a familiar, user-friendly API
+- Surface enough control for full `astro` parity, but abstract away internal APIs
+- Enable third-party tools to consume `.astro` files on the server
+- Enable unit testing of `.astro` component output
+- Possibly unblock `.mdx` compiledContent/html output?
+- Support Astro framework renderers (`@astrojs/*`) if possible
+
+# Non-Goals
+
+- Provide a way to **import** `.astro` components. Users will be responsible to provide a compiled component.
+- Provide a way to **import** `astro.config.(mjs|mts)` out of the box.
+
+# Detailed Design
+
+## Preface
+
+**This** RFC will have a smaller scoped compared to what users have envisioned. The reason why the scope shrunk is that 
+I want to ship a smaller feature on the surface, which will allow us, later to enhance it based on user's feedback and use cases.
```

Comment:
There's a mix of I / We in the RFC, even this sentence include both "**I** want" and "allow **us**". I personally believe it's best to always use `we` when we're writing RFCs from the perspective of the core team, but either works as long as it's consistent!